File: LC187RepeatedDNASequences.java
package leetcode;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;

/**
 * https://leetcode.com/problems/repeated-dna-sequences/
 */
public class LC187RepeatedDNASequences {
    public class Solution {
        public int compress(char nucleotide) {
            switch (nucleotide) {
                case 'A':
                    return 0;
                case 'C':
                    return 1;
                case 'G':
                    return 2;
                case 'T':
                    return 3;
            }
            return -1;
        }

        public int encode(int sequence, int compressed) {
            return (compressed << 18) | (sequence >> 2);
        }

        public String decode(int sequence) {
            StringBuilder decoded = new StringBuilder();
            for (int i = 0; i < 10; i++) {
                int nucleotide = sequence & 0x3;
                switch (nucleotide) {
                    case 0:
                        decoded.append('A');
                        break;
                    case 1:
                        decoded.append('C');
                        break;
                    case 2:
                        decoded.append('G');
                        break;
                    case 3:
                        decoded.append('T');
                        break;
                }
                sequence >>= 2;
            }
            return decoded.toString();
        }

        public List<String> findRepeatedDnaSequences(String s) {
            List<String> list = new ArrayList<>();
            if (s == null || s.length() < 10) {
                return list;
            }
            Set<Integer> seen = new HashSet<>();
            int now = 0;
            for (int i = 0; i < 9; i++) {
                now = encode(now, compress(s.charAt(i)));
            }
            for (int i = 9; i < s.length(); i++) {
                now = encode(now, compress(s.charAt(i)));
                if (seen.contains(now | 0xFFF00000)) {
                    continue;
                }
                if (seen.contains(now)) {
                    seen.remove(now);
                    seen.add(now | 0xFFF00000);
                    continue;
                }
                seen.add(now);
            }
            for (int sequence : seen) {
                if (sequence < 0) {
                    list.add(decode(sequence & 0xFFFFF));
                }
            }
            return list;
        }
    }

    @Test
    public void test_AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT() throws Exception {
        String[] expected = {"AAAAACCCCC", "CCCCCAAAAA"};
        Object[] result = new Solution().findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT").toArray();
        Arrays.sort(result);
        assertArrayEquals(expected, result);
    }
}

File: algorithms/code/leetcode/lc187_repeated_dna_sequences/lc187_repeated_dna_sequences.rs
// https://leetcode.com/problems/repeated-dna-sequences/
// #medium

use std::collections::HashSet;

pub struct Solution;

impl Solution {
    fn compress(nucleotide: &str) -> usize {
        match nucleotide.chars().nth(0) {
            Some('A') => 0,
            Some('C') => 1,
            Some('G') => 2,
            _ => 3,
        }
    }

    fn encode(sequence: u64, compressed: usize) -> u64 {
        ((compressed as u64) << 18) | (sequence >> 2)
    }

    fn decode(sequence_value: u64) -> String {
        let lut_to_int = ["A", "C", "G", "T"];
        let mut sequence = sequence_value;
        let mut decoded = String::new();
        for _ in 0..10 {
            let nucleotide = (sequence & 0x3) as usize;
            decoded.push_str(lut_to_int[nucleotide]);
            sequence >>= 2;
        }
        decoded
    }

    pub fn find_repeated_dna_sequences(s: &str) -> Vec<String> {
        let mut list = Vec::new();
        if s.len() < 10 {
            return list;
        }
        let mut seen = HashSet::new();
        let mut now = 0;
        for i in 0..9 {
            now = Self::encode(now, Self::compress(&s[i..i + 1]));
        }
        for i in 9..s.len() {
            now = Self::encode(now, Self::compress(&s[i..i + 1]));
            if seen.contains(&(now | 0xFFF00000)) {
                continue;
            }
            if seen.contains(&now) {
                seen.remove(&now);
                seen.insert(now | 0xFFF00000);
                continue;
            }
            seen.insert(now);
        }
        for sequence in seen {
            if sequence & 0xFFF00000 == 0xFFF00000 {
                list.push(Self::decode(sequence));
            }
        }
        list
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_aaaaacccccaaaaaccccccaaaaagggttt() {
        let expected = vec!["AAAAACCCCC".to_string(), "CCCCCAAAAA".to_string()];
        let mut result = Solution::find_repeated_dna_sequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT");
        result.sort_unstable();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_nothing() {
        let expected: Vec<String> = vec![];
        let result = Solution::find_repeated_dna_sequences("");
        assert_eq!(result, expected);
    }

    #[test]
    fn test_aaaaaaaaaaaaa() {
        let expected: Vec<String> = vec!["AAAAAAAAAA".to_string()];
        let mut result = Solution::find_repeated_dna_sequences("AAAAAAAAAAAAA");
        result.sort_unstable();
        assert_eq!(result, expected);
    }
}
